// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: model/common.proto

package model

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ErrorResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ErrorResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ErrorResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ErrorResponseMultiError, or
// nil if none found.
func (m *ErrorResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ErrorResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for CorrelationId

	if all {
		switch v := interface{}(m.GetError()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ErrorResponseValidationError{
					field:  "Error",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ErrorResponseValidationError{
					field:  "Error",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetError()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ErrorResponseValidationError{
				field:  "Error",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ErrorResponseMultiError(errors)
	}

	return nil
}

// ErrorResponseMultiError is an error wrapping multiple validation errors
// returned by ErrorResponse.ValidateAll() if the designated constraints
// aren't met.
type ErrorResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ErrorResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ErrorResponseMultiError) AllErrors() []error { return m }

// ErrorResponseValidationError is the validation error returned by
// ErrorResponse.Validate if the designated constraints aren't met.
type ErrorResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ErrorResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ErrorResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ErrorResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ErrorResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ErrorResponseValidationError) ErrorName() string { return "ErrorResponseValidationError" }

// Error satisfies the builtin error interface
func (e ErrorResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sErrorResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ErrorResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ErrorResponseValidationError{}

// Validate checks the field values on ErrorObject with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ErrorObject) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ErrorObject with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ErrorObjectMultiError, or
// nil if none found.
func (m *ErrorObject) ValidateAll() error {
	return m.validate(true)
}

func (m *ErrorObject) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for Message

	if len(errors) > 0 {
		return ErrorObjectMultiError(errors)
	}

	return nil
}

// ErrorObjectMultiError is an error wrapping multiple validation errors
// returned by ErrorObject.ValidateAll() if the designated constraints aren't met.
type ErrorObjectMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ErrorObjectMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ErrorObjectMultiError) AllErrors() []error { return m }

// ErrorObjectValidationError is the validation error returned by
// ErrorObject.Validate if the designated constraints aren't met.
type ErrorObjectValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ErrorObjectValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ErrorObjectValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ErrorObjectValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ErrorObjectValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ErrorObjectValidationError) ErrorName() string { return "ErrorObjectValidationError" }

// Error satisfies the builtin error interface
func (e ErrorObjectValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sErrorObject.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ErrorObjectValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ErrorObjectValidationError{}

// Validate checks the field values on GenrateTokenRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GenrateTokenRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GenrateTokenRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GenrateTokenRequestMultiError, or nil if none found.
func (m *GenrateTokenRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GenrateTokenRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if !_GenrateTokenRequest_MagicCode_Pattern.MatchString(m.GetMagicCode()) {
		err := GenrateTokenRequestValidationError{
			field:  "MagicCode",
			reason: "value does not match regex pattern \"^[a-zA-Z0-9]{1, 32}$\"",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if !_GenrateTokenRequest_Qrcode_Pattern.MatchString(m.GetQrcode()) {
		err := GenrateTokenRequestValidationError{
			field:  "Qrcode",
			reason: "value does not match regex pattern \"^[a-zA-Z0-9]{1, 32}$\"",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return GenrateTokenRequestMultiError(errors)
	}

	return nil
}

// GenrateTokenRequestMultiError is an error wrapping multiple validation
// errors returned by GenrateTokenRequest.ValidateAll() if the designated
// constraints aren't met.
type GenrateTokenRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GenrateTokenRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GenrateTokenRequestMultiError) AllErrors() []error { return m }

// GenrateTokenRequestValidationError is the validation error returned by
// GenrateTokenRequest.Validate if the designated constraints aren't met.
type GenrateTokenRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GenrateTokenRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GenrateTokenRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GenrateTokenRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GenrateTokenRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GenrateTokenRequestValidationError) ErrorName() string {
	return "GenrateTokenRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GenrateTokenRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGenrateTokenRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GenrateTokenRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GenrateTokenRequestValidationError{}

var _GenrateTokenRequest_MagicCode_Pattern = regexp.MustCompile("^[a-zA-Z0-9]{1, 32}$")

var _GenrateTokenRequest_Qrcode_Pattern = regexp.MustCompile("^[a-zA-Z0-9]{1, 32}$")

// Validate checks the field values on GenrateTokenResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GenrateTokenResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GenrateTokenResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GenrateTokenResponseMultiError, or nil if none found.
func (m *GenrateTokenResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GenrateTokenResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Token

	// no validation rules for Type

	// no validation rules for Expiry

	if len(errors) > 0 {
		return GenrateTokenResponseMultiError(errors)
	}

	return nil
}

// GenrateTokenResponseMultiError is an error wrapping multiple validation
// errors returned by GenrateTokenResponse.ValidateAll() if the designated
// constraints aren't met.
type GenrateTokenResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GenrateTokenResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GenrateTokenResponseMultiError) AllErrors() []error { return m }

// GenrateTokenResponseValidationError is the validation error returned by
// GenrateTokenResponse.Validate if the designated constraints aren't met.
type GenrateTokenResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GenrateTokenResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GenrateTokenResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GenrateTokenResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GenrateTokenResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GenrateTokenResponseValidationError) ErrorName() string {
	return "GenrateTokenResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GenrateTokenResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGenrateTokenResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GenrateTokenResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GenrateTokenResponseValidationError{}

// Validate checks the field values on GenerateApiKeyResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GenerateApiKeyResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GenerateApiKeyResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GenerateApiKeyResponseMultiError, or nil if none found.
func (m *GenerateApiKeyResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GenerateApiKeyResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ApiKey

	if len(errors) > 0 {
		return GenerateApiKeyResponseMultiError(errors)
	}

	return nil
}

// GenerateApiKeyResponseMultiError is an error wrapping multiple validation
// errors returned by GenerateApiKeyResponse.ValidateAll() if the designated
// constraints aren't met.
type GenerateApiKeyResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GenerateApiKeyResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GenerateApiKeyResponseMultiError) AllErrors() []error { return m }

// GenerateApiKeyResponseValidationError is the validation error returned by
// GenerateApiKeyResponse.Validate if the designated constraints aren't met.
type GenerateApiKeyResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GenerateApiKeyResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GenerateApiKeyResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GenerateApiKeyResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GenerateApiKeyResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GenerateApiKeyResponseValidationError) ErrorName() string {
	return "GenerateApiKeyResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GenerateApiKeyResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGenerateApiKeyResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GenerateApiKeyResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GenerateApiKeyResponseValidationError{}

// Validate checks the field values on OAuthCallbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *OAuthCallbackRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuthCallbackRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// OAuthCallbackRequestMultiError, or nil if none found.
func (m *OAuthCallbackRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuthCallbackRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Code

	// no validation rules for State

	if len(errors) > 0 {
		return OAuthCallbackRequestMultiError(errors)
	}

	return nil
}

// OAuthCallbackRequestMultiError is an error wrapping multiple validation
// errors returned by OAuthCallbackRequest.ValidateAll() if the designated
// constraints aren't met.
type OAuthCallbackRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuthCallbackRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuthCallbackRequestMultiError) AllErrors() []error { return m }

// OAuthCallbackRequestValidationError is the validation error returned by
// OAuthCallbackRequest.Validate if the designated constraints aren't met.
type OAuthCallbackRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuthCallbackRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuthCallbackRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuthCallbackRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuthCallbackRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuthCallbackRequestValidationError) ErrorName() string {
	return "OAuthCallbackRequestValidationError"
}

// Error satisfies the builtin error interface
func (e OAuthCallbackRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuthCallbackRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuthCallbackRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuthCallbackRequestValidationError{}

// Validate checks the field values on OAuthURLRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *OAuthURLRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuthURLRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// OAuthURLRequestMultiError, or nil if none found.
func (m *OAuthURLRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuthURLRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return OAuthURLRequestMultiError(errors)
	}

	return nil
}

// OAuthURLRequestMultiError is an error wrapping multiple validation errors
// returned by OAuthURLRequest.ValidateAll() if the designated constraints
// aren't met.
type OAuthURLRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuthURLRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuthURLRequestMultiError) AllErrors() []error { return m }

// OAuthURLRequestValidationError is the validation error returned by
// OAuthURLRequest.Validate if the designated constraints aren't met.
type OAuthURLRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuthURLRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuthURLRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuthURLRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuthURLRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuthURLRequestValidationError) ErrorName() string { return "OAuthURLRequestValidationError" }

// Error satisfies the builtin error interface
func (e OAuthURLRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuthURLRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuthURLRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuthURLRequestValidationError{}

// Validate checks the field values on OAuthProvider with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *OAuthProvider) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuthProvider with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in OAuthProviderMultiError, or
// nil if none found.
func (m *OAuthProvider) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuthProvider) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	// no validation rules for Url

	if len(errors) > 0 {
		return OAuthProviderMultiError(errors)
	}

	return nil
}

// OAuthProviderMultiError is an error wrapping multiple validation errors
// returned by OAuthProvider.ValidateAll() if the designated constraints
// aren't met.
type OAuthProviderMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuthProviderMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuthProviderMultiError) AllErrors() []error { return m }

// OAuthProviderValidationError is the validation error returned by
// OAuthProvider.Validate if the designated constraints aren't met.
type OAuthProviderValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuthProviderValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuthProviderValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuthProviderValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuthProviderValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuthProviderValidationError) ErrorName() string { return "OAuthProviderValidationError" }

// Error satisfies the builtin error interface
func (e OAuthProviderValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuthProvider.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuthProviderValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuthProviderValidationError{}

// Validate checks the field values on OAuthURLResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *OAuthURLResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OAuthURLResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// OAuthURLResponseMultiError, or nil if none found.
func (m *OAuthURLResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *OAuthURLResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetProviders() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, OAuthURLResponseValidationError{
						field:  fmt.Sprintf("Providers[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, OAuthURLResponseValidationError{
						field:  fmt.Sprintf("Providers[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return OAuthURLResponseValidationError{
					field:  fmt.Sprintf("Providers[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return OAuthURLResponseMultiError(errors)
	}

	return nil
}

// OAuthURLResponseMultiError is an error wrapping multiple validation errors
// returned by OAuthURLResponse.ValidateAll() if the designated constraints
// aren't met.
type OAuthURLResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OAuthURLResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OAuthURLResponseMultiError) AllErrors() []error { return m }

// OAuthURLResponseValidationError is the validation error returned by
// OAuthURLResponse.Validate if the designated constraints aren't met.
type OAuthURLResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OAuthURLResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OAuthURLResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OAuthURLResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OAuthURLResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OAuthURLResponseValidationError) ErrorName() string { return "OAuthURLResponseValidationError" }

// Error satisfies the builtin error interface
func (e OAuthURLResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOAuthURLResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OAuthURLResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OAuthURLResponseValidationError{}

// Validate checks the field values on GenerateWechatQRCodeResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GenerateWechatQRCodeResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GenerateWechatQRCodeResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GenerateWechatQRCodeResponseMultiError, or nil if none found.
func (m *GenerateWechatQRCodeResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GenerateWechatQRCodeResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Url

	// no validation rules for Qrcode

	if len(errors) > 0 {
		return GenerateWechatQRCodeResponseMultiError(errors)
	}

	return nil
}

// GenerateWechatQRCodeResponseMultiError is an error wrapping multiple
// validation errors returned by GenerateWechatQRCodeResponse.ValidateAll() if
// the designated constraints aren't met.
type GenerateWechatQRCodeResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GenerateWechatQRCodeResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GenerateWechatQRCodeResponseMultiError) AllErrors() []error { return m }

// GenerateWechatQRCodeResponseValidationError is the validation error returned
// by GenerateWechatQRCodeResponse.Validate if the designated constraints
// aren't met.
type GenerateWechatQRCodeResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GenerateWechatQRCodeResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GenerateWechatQRCodeResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GenerateWechatQRCodeResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GenerateWechatQRCodeResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GenerateWechatQRCodeResponseValidationError) ErrorName() string {
	return "GenerateWechatQRCodeResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GenerateWechatQRCodeResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGenerateWechatQRCodeResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GenerateWechatQRCodeResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GenerateWechatQRCodeResponseValidationError{}

// Validate checks the field values on Prompt with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Prompt) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Prompt with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in PromptMultiError, or nil if none found.
func (m *Prompt) ValidateAll() error {
	return m.validate(true)
}

func (m *Prompt) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Act

	// no validation rules for Prompt

	// no validation rules for IsCustom

	if len(errors) > 0 {
		return PromptMultiError(errors)
	}

	return nil
}

// PromptMultiError is an error wrapping multiple validation errors returned by
// Prompt.ValidateAll() if the designated constraints aren't met.
type PromptMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PromptMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PromptMultiError) AllErrors() []error { return m }

// PromptValidationError is the validation error returned by Prompt.Validate if
// the designated constraints aren't met.
type PromptValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PromptValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PromptValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PromptValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PromptValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PromptValidationError) ErrorName() string { return "PromptValidationError" }

// Error satisfies the builtin error interface
func (e PromptValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPrompt.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PromptValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PromptValidationError{}

// Validate checks the field values on ListPromptsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListPromptsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListPromptsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListPromptsRequestMultiError, or nil if none found.
func (m *ListPromptsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListPromptsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for IsCustom

	if len(errors) > 0 {
		return ListPromptsRequestMultiError(errors)
	}

	return nil
}

// ListPromptsRequestMultiError is an error wrapping multiple validation errors
// returned by ListPromptsRequest.ValidateAll() if the designated constraints
// aren't met.
type ListPromptsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListPromptsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListPromptsRequestMultiError) AllErrors() []error { return m }

// ListPromptsRequestValidationError is the validation error returned by
// ListPromptsRequest.Validate if the designated constraints aren't met.
type ListPromptsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListPromptsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListPromptsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListPromptsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListPromptsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListPromptsRequestValidationError) ErrorName() string {
	return "ListPromptsRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListPromptsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListPromptsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListPromptsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListPromptsRequestValidationError{}

// Validate checks the field values on ListPromptsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListPromptsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListPromptsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListPromptsResponseMultiError, or nil if none found.
func (m *ListPromptsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListPromptsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetPrompts() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListPromptsResponseValidationError{
						field:  fmt.Sprintf("Prompts[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListPromptsResponseValidationError{
						field:  fmt.Sprintf("Prompts[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListPromptsResponseValidationError{
					field:  fmt.Sprintf("Prompts[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListPromptsResponseMultiError(errors)
	}

	return nil
}

// ListPromptsResponseMultiError is an error wrapping multiple validation
// errors returned by ListPromptsResponse.ValidateAll() if the designated
// constraints aren't met.
type ListPromptsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListPromptsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListPromptsResponseMultiError) AllErrors() []error { return m }

// ListPromptsResponseValidationError is the validation error returned by
// ListPromptsResponse.Validate if the designated constraints aren't met.
type ListPromptsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListPromptsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListPromptsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListPromptsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListPromptsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListPromptsResponseValidationError) ErrorName() string {
	return "ListPromptsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListPromptsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListPromptsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListPromptsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListPromptsResponseValidationError{}

// Validate checks the field values on GetPromptRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetPromptRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetPromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetPromptRequestMultiError, or nil if none found.
func (m *GetPromptRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetPromptRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetPromptRequestMultiError(errors)
	}

	return nil
}

// GetPromptRequestMultiError is an error wrapping multiple validation errors
// returned by GetPromptRequest.ValidateAll() if the designated constraints
// aren't met.
type GetPromptRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetPromptRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetPromptRequestMultiError) AllErrors() []error { return m }

// GetPromptRequestValidationError is the validation error returned by
// GetPromptRequest.Validate if the designated constraints aren't met.
type GetPromptRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetPromptRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetPromptRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetPromptRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetPromptRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetPromptRequestValidationError) ErrorName() string { return "GetPromptRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetPromptRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetPromptRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetPromptRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetPromptRequestValidationError{}

// Validate checks the field values on GetPromptResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetPromptResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetPromptResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetPromptResponseMultiError, or nil if none found.
func (m *GetPromptResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetPromptResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetPrompt()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetPromptResponseValidationError{
					field:  "Prompt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetPromptResponseValidationError{
					field:  "Prompt",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetPrompt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetPromptResponseValidationError{
				field:  "Prompt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetPromptResponseMultiError(errors)
	}

	return nil
}

// GetPromptResponseMultiError is an error wrapping multiple validation errors
// returned by GetPromptResponse.ValidateAll() if the designated constraints
// aren't met.
type GetPromptResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetPromptResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetPromptResponseMultiError) AllErrors() []error { return m }

// GetPromptResponseValidationError is the validation error returned by
// GetPromptResponse.Validate if the designated constraints aren't met.
type GetPromptResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetPromptResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetPromptResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetPromptResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetPromptResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetPromptResponseValidationError) ErrorName() string {
	return "GetPromptResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetPromptResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetPromptResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetPromptResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetPromptResponseValidationError{}

// Validate checks the field values on CreatePromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreatePromptRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreatePromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreatePromptRequestMultiError, or nil if none found.
func (m *CreatePromptRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreatePromptRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Act

	// no validation rules for Prompt

	if len(errors) > 0 {
		return CreatePromptRequestMultiError(errors)
	}

	return nil
}

// CreatePromptRequestMultiError is an error wrapping multiple validation
// errors returned by CreatePromptRequest.ValidateAll() if the designated
// constraints aren't met.
type CreatePromptRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreatePromptRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreatePromptRequestMultiError) AllErrors() []error { return m }

// CreatePromptRequestValidationError is the validation error returned by
// CreatePromptRequest.Validate if the designated constraints aren't met.
type CreatePromptRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreatePromptRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreatePromptRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreatePromptRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreatePromptRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreatePromptRequestValidationError) ErrorName() string {
	return "CreatePromptRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreatePromptRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreatePromptRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreatePromptRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreatePromptRequestValidationError{}

// Validate checks the field values on DeletePromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeletePromptRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeletePromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeletePromptRequestMultiError, or nil if none found.
func (m *DeletePromptRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeletePromptRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return DeletePromptRequestMultiError(errors)
	}

	return nil
}

// DeletePromptRequestMultiError is an error wrapping multiple validation
// errors returned by DeletePromptRequest.ValidateAll() if the designated
// constraints aren't met.
type DeletePromptRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeletePromptRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeletePromptRequestMultiError) AllErrors() []error { return m }

// DeletePromptRequestValidationError is the validation error returned by
// DeletePromptRequest.Validate if the designated constraints aren't met.
type DeletePromptRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeletePromptRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeletePromptRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeletePromptRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeletePromptRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeletePromptRequestValidationError) ErrorName() string {
	return "DeletePromptRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeletePromptRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeletePromptRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeletePromptRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeletePromptRequestValidationError{}

// Validate checks the field values on UpdatePromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdatePromptRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdatePromptRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdatePromptRequestMultiError, or nil if none found.
func (m *UpdatePromptRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdatePromptRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Act

	// no validation rules for Prompt

	if len(errors) > 0 {
		return UpdatePromptRequestMultiError(errors)
	}

	return nil
}

// UpdatePromptRequestMultiError is an error wrapping multiple validation
// errors returned by UpdatePromptRequest.ValidateAll() if the designated
// constraints aren't met.
type UpdatePromptRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdatePromptRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdatePromptRequestMultiError) AllErrors() []error { return m }

// UpdatePromptRequestValidationError is the validation error returned by
// UpdatePromptRequest.Validate if the designated constraints aren't met.
type UpdatePromptRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdatePromptRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdatePromptRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdatePromptRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdatePromptRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdatePromptRequestValidationError) ErrorName() string {
	return "UpdatePromptRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdatePromptRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdatePromptRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdatePromptRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdatePromptRequestValidationError{}
